Ability to buy tokens with Moonpay

nikoferro opened the pull request:
## Explanation

**This extends this already merged PR**: https://github.com/MetaMask/metamask-extension/pull/15551
and adds Moonpay as an extra onramp provider

Current State: Only native currencies can be bought through Moonpay using the buy button
After Change: Ability to purchase tokens defined in BUYABLE_CHAINS_MAP using Moonpay

<!--
Thanks for the pull request. Take a moment to answer these questions so that reviewers have the information they need to properly understand your changes:

* What is the current state of things and why does it need to change?
* What is the solution your changes offer and how does it work?

Below is a template to give you some ideas. Feel free to use your own words!

Currently, ...

This is a problem because ...

In order to solve this problem, this pull request ...
-->

## More Information

See ONRAMP-189 and ONRAMP-190

<!--
Are there any issues, Slack conversations, Zendesk issues, user stories, etc. reviewers should consult to understand this pull request better? For instance:

* Fixes #12345
* See: #67890
-->

## Screenshots/Screencaps

<!-- If you're making a change to the UI, make sure to capture a screenshot or a short video showing off your work! -->

### Before
<img width="354" alt="Screenshot 2022-09-21 at 16 12 46" src="https://user-images.githubusercontent.com/8658960/191527879-38353149-f7b4-4a40-a6db-e7d9497a5e80.png">

<!-- How did the UI you changed look before your changes? Drag your file(s) below this line: -->

### After
<img width="355" alt="Screenshot 2022-09-21 at 16 12 32" src="https://user-images.githubusercontent.com/8658960/191527910-9d3fa242-616e-446a-988d-20f866931745.png">

<!-- How does it look now? Drag your file(s) below this line: -->

## Manual Testing Steps

These changes apply only to Ethereum mainnet

Import a token
If token is on the Moonpay lists it should display a "Buy" button

Moonpay lists:

[Mainnet](https://github.com/nikoferro/metamask-extension/commit/15959485c202fa0dac3fd1b4f0b0d934a44c80a4#diff-4678018988a2b2341bb3addb3e8ae051c9fe7778e7b8ed3fdd7834a3f5912dd2R683)
[BSC](https://github.com/nikoferro/metamask-extension/commit/15959485c202fa0dac3fd1b4f0b0d934a44c80a4#diff-4678018988a2b2341bb3addb3e8ae051c9fe7778e7b8ed3fdd7834a3f5912dd2R828):
[Polygon](https://github.com/nikoferro/metamask-extension/commit/15959485c202fa0dac3fd1b4f0b0d934a44c80a4#diff-4678018988a2b2341bb3addb3e8ae051c9fe7778e7b8ed3fdd7834a3f5912dd2R845)
[Avalanche C-Chain](https://github.com/nikoferro/metamask-extension/commit/15959485c202fa0dac3fd1b4f0b0d934a44c80a4#diff-4678018988a2b2341bb3addb3e8ae051c9fe7778e7b8ed3fdd7834a3f5912dd2R859)
[Celo](https://github.com/nikoferro/metamask-extension/commit/15959485c202fa0dac3fd1b4f0b0d934a44c80a4#diff-4678018988a2b2341bb3addb3e8ae051c9fe7778e7b8ed3fdd7834a3f5912dd2R880)

Depending on which list it shows, it should give you the option to buy the token on Moonpay
Once you click on the provider's "Continue to" button, a window on said provider should open, and the token to buy should already be preselected

## Pre-Merge Checklist

- [x] PR template is filled out
- [ ] **IF** this PR fixes a bug, a test that _would have_ caught the bug has been added
- [ ] PR is linked to the appropriate GitHub issue
- [ ] PR has been added to the appropriate release Milestone

### + If there are functional changes:

- [ ] Manual testing complete & passed
- [ ] "Extension QA Board" label has been applied

nikoferro committed:
Ability to buy tokens with Moonpay

nikoferro committed:
Merge branch 'develop' into purchase-moonpay

nikoferro commented:
@brad-decker tagging you here since you also went through the previous [PR](https://github.com/MetaMask/metamask-extension/pull/15551) and you are already familiar with these changes

brad-decker commented:
@nikoferro did this pass locally for you? I have a feeling that the "type is too complex" issue will come back here due to how long supported currencies list is.

nikoferro committed:
fix for test cases failing

nikoferro committed:
Merge branch 'purchase-moonpay' of github.com:nikoferro/metamask-extension into purchase-moonpay

nikoferro commented:
@brad-decker yes, worked fine on my end, and seems to be passing tests now. had to do a couple of changes on how we handle supported currencies for moonpay.

i.e: triple checked and symbols are not actually required to be lowercase

brad-decker commented on shared/constants/network.ts:
```diff
@@ -49,25 +44,21 @@ type MoonPayNetworkAbbreviation = 'bsc' | 'cchain' | 'polygon';
-    | `${LowercaseCurrencySymbol},${LowercaseCurrencySymbol},${LowercaseCurrencySymbol},${LowercaseCurrencySymbol}`
-    | `${LowercaseCurrencySymbol}_${MoonPayNetworkAbbreviation}`
-    | `${LowercaseCurrencySymbol}_${MoonPayNetworkAbbreviation},${LowercaseCurrencySymbol}_${MoonPayNetworkAbbreviation}`;
+    | SupportedCurrencySymbol[]
+    | `${SupportedCurrencySymbol}_${MoonPayNetworkAbbreviation}`[];
```
Oh i see its an array now instead of a joined list. Could you update the doc block above to remove the '1 to 3' joined by ',' and '1 or 2' 'and concatenated with ','' portions? The array is much better and the reason why this is passing TSC now 👏🏻

nikoferro replied:
thanks for taking a look @brad-decker ! good catch btw, just updated the doc block that was describing the previous typing.

brad-decker reviewed (dismissed):
This is all good except a doc change request, i'm approving emphatically but if you have time to fixup that doc i'll reapprove after the commit ;P

nikoferro committed:
updated description for MoonPayChainSettings type

nikoferro committed:
Merge branch 'develop' into purchase-moonpay

brad-decker commented:
Looks like junit test results got commited somehow

nikoferro committed:
removed test results

nikoferro commented:
> Looks like junit test results got commited somehow

@brad-decker yeah i noticed that file, but since it wasn't on the gitignore, i understood that it was intended to be included (maybe just for an extra check, idk)

removed the results and added `test-results/` to the gitignore as i get from your comment that this was not intended to be included

nikoferro committed:
Merge branch 'purchase-moonpay' of github.com:nikoferro/metamask-extension into purchase-moonpay

ryanml reviewed (approved):
lgtm
